lib/pure/hashes.nim

```diff
@@ -86,6 +89,50 @@ proc hashData*(data: pointer, size: int): Hash =
     dec(s)
   result = !$h
 
+proc hashBiggestIntVM(x: BiggestInt): Hash = discard # in vmops
+
+proc hash*(x: string): Hash
+
+proc hashBiggestInt*(x: BiggestInt): Hash {.inline.} =
+  ## for internal use; user code should prefer `hash` overloads
+  when nimvm: hashBiggestIntVM(x)
+  else:
+    when defined(js):
+      # could use BigInt, see https://github.com/nim-lang/RFCs/issues/187
+      hash($x)
+    else:
+      hashData(cast[pointer](unsafeAddr x), type(x).sizeof)
+
+proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} =
+  ## Efficient hashing of numbers, ordinals (eg enum), char.
+  when T.sizeof >= 4:
```

Comment:
I don't entirely understand why this is even needed. Hash is an int, so casting to int should just do it no?

Reply:
no: with your proposal `proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} = cast[Hash](x)` you get bad results:
* hash(0.0) != hash(-0.0)
* hash(x) is very non random
* hash(x) is doesn't match between RT and CT
```
     0.3 RT => 4599075939470750515
     0.3 CT => 4599075939470750515
     0.0 RT => 0
     0.0 CT => 0
    -0.0 RT => -9223372036854775808
    -0.0 CT => 0
   "foo" RT => 4138058784
   "foo" CT => 4138058784
     123 RT => 123
     123 CT => 123
     124 RT => 124
     124 CT => 124
     125 RT => 125
```

with the algo I have you get:
```
     0.3 RT => -7272369355435918195
     0.3 CT => -7272369355435918195
     0.0 RT => 0
     0.0 CT => 0
    -0.0 RT => 0
    -0.0 CT => 0
   "foo" RT => 4138058784
   "foo" CT => 4138058784
     123 RT => -508526132817443961
     123 CT => -508526132817443961
     124 RT => -1530243182942418667
     124 CT => -1530243182942418667
     125 RT => -6335688243927128111
     125 CT => -6335688243927128111
```

Reply:
> hash(x) is very non random
It doesn't need to be random when its perfect.
> hash(0.0) != hash(-0.0)
As expected? Floats work this way. Not the hashsets problem IMO
> hash(x) is doesn't match between RT and CT
That can be fixed

Reply:
> It doesn't need to be random when its perfect.

this "perfect" identity hash of yours is the exact reason you get slow insertions in https://github.com/nim-lang/Nim/issues/13393; this can result in slowdowns of 100x-1000x, see tests in top post and in that issue

>> hash(0.0) != hash(-0.0)

> As expected? Floats work this way. Not the hashsets problem IMO

sorry, no. I don't know any language where `x == y` is true but hash(x) != hash(y). This makes no mathematical/logical/practical sense.


> > hash(x) is doesn't match between RT and CT

> That can be fixed

indeed; this PR doesn't have this problem

Reply:
> this "perfect" identity hash of yours is the exact reason you get slow insertions in #13393; this can result in slowdowns of 100x-1000x, see tests in top post and in that issue
No you are wrong: https://github.com/nim-lang/Nim/pull/13410/files#diff-466aee2707291e60fe359bda6b675b86L118
> sorry, no. I don't know any language where x == y is true but hash(x) != hash(y). This makes no mathematical/logical/practical sense.
Oh, I wasn't aware that `0.0 == -0.0` is true.
Ok so we do it your way for floats.

Reply:
> > this "perfect" identity hash of yours is the exact reason you get slow insertions in #13393; this can result in slowdowns of 100x-1000x, see tests in top post and in that issue

> No you are wrong: nim-lang/Nim/pull/13410/files#diff-466aee2707291e60fe359bda6b675b86L118

not sure where I'm wrong; I was showing what happened if I followed your suggestion `I don't entirely understand why this is even needed. Hash is an int, so casting to int should just do it no?`, so I showed results above for
`proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} = cast[Hash](x)`

this results in predictable, non-random values that cause massive slowdowns during insertions, see benchmarks in https://github.com/nim-lang/Nim/issues/13393 (run it to see for yourself!)

using a prime as was done in nim-lang/Nim/pull/13410/files#diff-466aee2707291e60fe359bda6b675b86L118 via `cast[Hash](cast[uint](x) * prime)` didn't properly solve this problem, as evidenced by that benchmark.

Using this PR, you get a 100x-1000x speedup on this benchmark (which I didn't write btw).

Reply:
Sorry, I'm stupid :D